synapse/handlers/presence.py

```diff
@@ -1091,59 +1118,207 @@ async def get_new_events(
                 # doesn't return. C.f. #5503.
                 return [], max_token
 
-            presence = self.get_presence_handler()
-            stream_change_cache = self.store.presence_stream_cache
-
+            # Figure out which other users this user should receive updates for
             users_interested_in = await self._get_interested_in(user, explicit_room_id)
 
-            user_ids_changed = set()  # type: Collection[str]
-            changed = None
-            if from_key:
-                changed = stream_change_cache.get_all_entities_changed(from_key)
-
-            if changed is not None and len(changed) < 500:
-                assert isinstance(user_ids_changed, set)
+            # We have a set of users that we're interested in the presence of. We want to
+            # cross-reference that with the users that have actually changed their presence.
 
-                # For small deltas, its quicker to get all changes and then
-                # work out if we share a room or they're in our presence list
-                get_updates_counter.labels("stream").inc()
-                for other_user_id in changed:
-                    if other_user_id in users_interested_in:
-                        user_ids_changed.add(other_user_id)
-            else:
-                # Too many possible updates. Find all users we can see and check
-                # if any of them have changed.
-                get_updates_counter.labels("full").inc()
+            # Check whether this user should see all user updates
 
+            if users_interested_in == "ALL":
                 if from_key:
-                    user_ids_changed = stream_change_cache.get_entities_changed(
-                        users_interested_in, from_key
+                    # We need to return all new presence updates to this user, regardless of whether
+                    # they share a room with that user
+                    presence_updates = await self._filter_all_presence_updates_for_user(
+                        user_id, from_key, include_offline
                     )
+                    return presence_updates, max_token
                 else:
-                    user_ids_changed = users_interested_in
+                    # This user should receive all user presence, and hasn't provided a from_key.
+                    # Send all currently known user presence states.
+                    users_to_state = await self.store.get_presence_for_all_users(
+                        include_offline=include_offline
+                    )
 
-            updates = await presence.current_state_for_users(user_ids_changed)
+                    return list(users_to_state.values()), max_token
```

Comment:
Did we not decide that we wanted to filter users through `get_users_for_states` if we returned `ALL`?

Reply:
We did, and this is indeed a path that does not do so! I've made this path use `_filter_all_presence_updates_for_user` in 2f16ed047, as well as made that function a bit cleaner.